### PR TITLE
Environment Variables!

### LINKED ISSUE #12: Environment variables for commands
It would be useful to have some or all of the run-script build phase environment variables available to the executables that are run in response to changes.

### PR DESCRIPTION
**Pull request by vokal-isaac**
Resolves #12.

This provides the running task with the environment variables as reported by `xcodebuild -showBuildVariables`.  Because that's run synchronously, the task-launching has been moved to a background thread.

I wasn't sure how best to express this in the README, especially since using the environment variables will make the advice about testing the script in Terminal very hard to follow.

@designatednerd Wanna review this or should I `@`-ping the whole team?

### CODE REVIEW

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m [hunk 1]
```diff
@@ -116,6 +116,52 @@ - (VOKDirectoryWatcher *)directoryWatcher
     return _directoryWatcher;
 }
 
+- (NSDictionary *)environmentVariables
+{
+    if ([NSThread isMainThread]) {
+        NSLog(@"VOKProjectContainer environmentVariables being retrieved on the main thread.");
```
Comment (designatednerd) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: Maybe this ought to be an assert rather than a log?

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m [hunk 2]
```diff
@@ -116,6 +116,52 @@ - (VOKDirectoryWatcher *)directoryWatcher
...
+    buildSettingsTask.standardOutput = outputPipe;
+    
+    [buildSettingsTask launch];
+    [buildSettingsTask waitUntilExit];
+    
+    NSData *outputData = [outputPipe.fileHandleForReading readDataToEndOfFile];
+    NSString *outputString = [[NSString alloc] initWithData:outputData encoding:NSUTF8StringEncoding];
+    
+    NSError *error;
+    NSRegularExpression *regex = [NSRegularExpression
```
Comment (designatednerd) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: Probably worth adding a comment about what the shit this regex is doing since regex is, technically, pure voodoo

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m [hunk 3]
```diff
@@ -116,6 +116,52 @@ - (VOKDirectoryWatcher *)directoryWatcher
...
+        return nil;
+    }
+    
+    NSMutableDictionary *env = [NSMutableDictionary dictionary];
+    [regex
+     enumerateMatchesInString:outputString
+     options:NSMatchingReportProgress
+     range:NSMakeRange(0, outputString.length)
+     usingBlock:^(NSTextCheckingResult *result, NSMatchingFlags flags, BOOL *stop) {
+         if (result.numberOfRanges != 3) {
```
Comment (designatednerd) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: What does the 3 represent?

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m [hunk 4]
```diff
@@ -116,6 +116,52 @@ - (VOKDirectoryWatcher *)directoryWatcher
...
+             return;
+         }
+         id key = [outputString substringWithRange:[result rangeAtIndex:1]];
+         id value = [outputString substringWithRange:[result rangeAtIndex:2]];
```
Comment (designatednerd) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: What about 1 and 2?

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKScriptForFolder.m [hunk 5]
```diff
@@ -155,10 +155,13 @@ - (void)actuallyRunScript
...
     if ([[NSFileManager defaultManager] isExecutableFileAtPath:self.absolutePathToScript]) {
-        NSTask *task = [[NSTask alloc] init];
-        task.launchPath = self.absolutePathToScript;
-        task.currentDirectoryPath = [task.launchPath stringByDeletingLastPathComponent];
-        [task launch];
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+            NSTask *task = [[NSTask alloc] init];
+            task.launchPath = self.absolutePathToScript;
+            task.currentDirectoryPath = [task.launchPath stringByDeletingLastPathComponent];
+            task.environment = self.containingProject.environmentVariables;
```
Comment (designatednerd) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKScriptForFolder.m]: Hmmm - this will make every single task launched always run on a background thread, which might not be expected behavior. 

What about dispatching async to get the environment variables then dispatching back to the main queue to launch the task?

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m (hunk 1 above)
Comment (vokal-isaac) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: Maybe.  It was convenient for it not to assert during development, but I'm not sure I can justify it not asserting now... well, except that even if it asserts, the assert won't make it to release builds, but :shrugs:.

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m (hunk 2 above)
Comment (vokal-isaac) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: Regexes are not pure voodoo.  That said, how much detail?  Examples:
- `// Search for lines of the form [whitespace][key][whitespace]=[whitespace][value] and extract the key and value`
- http://stackoverflow.com/a/13827487/291280 and a note that this is an adaptation of that answer's regex
- http://rick.measham.id.au/paste/explain.pl?regex=%5E%5B%5B%3Aspace%3A%5D%5D*%28%5B%5E%5B%3Aspace%3A%5D%5D*%29%5B%5B%3Aspace%3A%5D%5D*%3D%5B%5B%3Aspace%3A%5D%5D*%28.*%29%24
- From https://regex101.com/:  
image

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m (hunk 3 above)
Comment (vokal-isaac) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: The 0th range is the range of the string that matches the whole regex; the 1st and 2nd ranges are the key and value; so there should be 3 ranges in the result—if there aren't, something's borked.

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m (hunk 4 above)
Comment (vokal-isaac) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: 1st and 2nd captures in the regex.

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKScriptForFolder.m (hunk 5 above)
Comment (vokal-isaac) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKScriptForFolder.m]: First, keep in mind that the main thread here is Xcode's UI thread.  Really, nothing that isn't touching the UI should be run on that thread.

The task was never being run on the main thread, since it's being launched asynchronously with `NSTask`, but it's also an independent process in a separate environment, so its execution is decoupled from Xcode's threads.  From the perspective of whatever's being run in the `NSTask`, there's no difference between creating and launching it on Xcode's UI thread or on a background thread, especially since we're not waiting for it to finish or doing anything to the UI as a result of its execution.

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m (hunk 1 above)
Comment (designatednerd) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: Might be useful at least for debug builds.

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m (hunk 2 above)
Comment (designatednerd) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: Option 2 please - when documenting regexes adapted from a particular source, link to the source if possible. 

And fine, they're not *pure* voodoo. Only like 85%.

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m (hunk 4 above)
Comment (designatednerd) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: I gathered that bit :stuck_out_tongue_closed_eyes: 

Maybe "key" and "value" just weren't descriptive enough for me - so the key is the environment variable and then the value is the full path it expands out to?

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m (hunk 3 above)
Comment (designatednerd) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: Probably worth leaving a comment along these lines for when regex dummies like me have to debug this.

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKScriptForFolder.m (hunk 5 above)
Comment (designatednerd) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKScriptForFolder.m]: Ahhhh, yeah, good point.

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m (hunk 2 above)
Comment (vokal-isaac) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: They are only voodoo when using lookarounds and stuff like that.

I'd say this is loosely based on that regex, at best, and follows from seeing that the output of `xcodebuild -showBuildSettings` is a list of lines of the form `[whitespace][key][whitespace]=[whitespace][value]`, but I'll put in whichever/whatever documenting comment you want here.

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m (hunk 3 above)
Comment (vokal-isaac) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: Essentially, it's a guard against going out of bounds when requesting the ranges at indexes 1 and 2, immediately below—that's why I didn't document it, but I'll add the comment.

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m (hunk 4 above)
Comment (vokal-isaac) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: The key is the variable name and the value is the value stored in that variable—it's not necessarily a path; it's an arbitrary value.  Hence the generic name.  That, and we're putting them into a dictionary as keys and values.

Comment (designatednerd) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: maybe "environmentVariable" instead of "key" - it's obvious we're using the env variable as a key, but it's not necessarily obvious that it is in fact the environment variable.

Comment (vokal-isaac) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: How about `variableName` and `variableValue`?

Comment (designatednerd) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: Works for me!

File: XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m (hunk 2 above)
Comment (vokal-isaac) [inline XcodeAutoBasher/XcodeAutoBasher/Models/VOKProjectContainer.m]: Comment added.

**Comment by designatednerd** [comment]
LGTM - @vokal/ios-developers Anyone else wanna have a look at this before we merge it down?

**Comment by bryanluby** [comment]
I'll take a look

**Comment by bryanluby** [comment]
LGTM
